src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/SimpleChunkVisitor.java

```diff
@@ -34,25 +34,42 @@
  * This visitor's callbacks are invoked as we walk through a pipeline flow graph, and it splits it into chunks.
  * <p> A {@link ForkScanner#visitSimpleChunks(SimpleChunkVisitor, ChunkFinder)} creates these FlowChunks using a {@link ChunkFinder} to define the chunk boundaries.
  *
- * <p> Implementations get to decide how to use and handle chunks.
- * <p> <h3>At a minimum they should handle:</h3>
+ * <p>We walk through the {@link FlowNode}s in reverse order from end to start, so <em>end callbacks are invoked before
+ *  their corresponding start callbacks.</em>
+ *
+ * <p><em>Callback types</em>
+ * <p> There are two kinds of callbacks - chunk callbacks, and parallel structure callbacks
+ * <p><em>Chunk Callbacks:</em>
  * <ul>
- *     <li>Unbalanced numbers of chunk start/end calls (for incomplete flows)</li>
- *     <li>A chunk end with no beginning (runs to start of flow, or never began)</li>
- *     <li>A chunk start with no end (ex: a block that hasn't completed running)</li>
- *     <li>Other starts/ends before we hit the closing one (nesting)</li>
- *     <li>Atom nodes not within the current Chunk (visitor is responsible for handling state)</li>
+ *     <li>{@link #chunkStart(FlowNode, FlowNode, ForkScanner)} - detected the start of a chunk beginning with a node</li>
+ *     <li>{@link #chunkEnd(FlowNode, FlowNode, ForkScanner)} - detected the end of a chunk, terminating with a node </li>
+ *     <li>{@link #atomNode(FlowNode, FlowNode, FlowNode, ForkScanner)} - most nodes, which aren't boundaries of chunks</li>
  * </ul>
  *
- * <em>Important implementation note: multiple callbacks can be invoked for a single node depending on its type.</em>
- * <p>For example, we may capture parallels as chunks.
+ * <p><p><em>Chunk callback rules:</em>
+ * <ol>
+ *     <li>For a single node, it may have EITHER OR BOTH chunkStart and chunkEnd events</li>
+ *     <li>Every node that doesn't get a startChunk/endChunk callback gets an atomNode callback.</li>
+ *     <li>For {@link ChunkFinder} implementations that match the {@link BlockStartNode} and {@link BlockEndNode} should never have both for a single</li>
+ *     <li>You cannot have multiple of any of the same specific type of callbacks for the same flownode</li>
+ *     <li>You cannot have a atomNode callback AND a start/end for the same flownode (application of the above).</li>
+ * </ol>
+ *
+ * <p>Parallel Structure Callbacks: Zero, One, or Multiple may be invoked for any given FlowNode</p>
```

Comment:
@vivek Exactly what I said -- there are special cases where a node may belong to more than one category (I've clarified in the latest commit)